alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,38 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
+fn shell_command(shell_config: Option<&Program>, pw: &Passwd<'_>) -> Command {
+    // Look for a shell in the alacritty config first, then SHELL, then in passwd.
+    let shell = shell_config
+        .map(|prog| prog.program().to_owned())
+        .or_else(|| env::var("SHELL").ok())
+        .unwrap_or_else(|| pw.shell.to_owned());
 
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
-}
+    #[cfg(not(target_os = "macos"))]
+    let mut command = Command::new(shell);
+
+    // On macOS, unless the user set the `shell` config in alacritty.yml, use the `login` command to
+    // launch the shell so programs like `w`, `who`, and `logname` work.
+    #[cfg(target_os = "macos")]
+    let mut command = if shell_config.is_some() {
+        Command::new(shell)
+    } else {
+        let mut login_command = Command::new("/usr/bin/login");
+        // -f bypasses authentication for the already-logged-in user
+        // -l makes it not a "login" shell and prevents login from changing directories to $HOME
+        // exec the shell with argv[0] prepended by - so the shell run as a login shell anyway
+        login_command.args(["-fl", pw.name, "/bin/zsh", "-c"]);
+        login_command.arg(format!("exec -a -{} {}", shell.rsplit('/').next().unwrap(), shell));
+        login_command
+    };
 
-#[cfg(not(target_os = "macos"))]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    Program::Just(env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned()))
+    // If the user specified a shell in alacritty.yml, the arguments are added here.
+    // Note that if we wanted to implicitly invoke these via login on macOS also, the arguments
+    // would need to be added to the `format!("exec ...")` above, but we only take that path if
+    // there is no shell config, and hence no arguments to add.
+    shell_config.map(|prog| command.args(prog.args()));
```

Comment:
What's the point of doing this when we know there's no shell config for one of the macos branches anyway?

Also `-e` can override the shell too, not just the config file. It also seems completely irrelevant what we *could* do here, if anything we should document reasons why we don't.

Reply:
This statement covers adding shell arguments for all platforms, and the comment is just explaining why the arguments can be added directly to the command. This will be obviated by separate functions.